src/idl_gen_csharp.cpp

```diff
@@ -1745,93 +1763,107 @@ class CSharpGenerator : public BaseGenerator {
     code += "  }\n";
   }
 
-  void GenStructArgs_ObjectAPI(const StructDef &struct_def,
-                               std::string *code_ptr,
-                               std::string prefix) const {
-    auto &code = *code_ptr;
-    for (auto it = struct_def.fields.vec.begin();
-         it != struct_def.fields.vec.end(); ++it) {
-      auto &field = **it;
-      const auto &field_type = field.value.type;
-      if (IsStruct(field_type)) {
-        GenStructArgs_ObjectAPI(*field_type.struct_def, code_ptr,
-                                prefix + "." + MakeCamel(field.name) + ".");
-      } else {
-        code += ",\n";
-        code += prefix + MakeCamel(field.name);
-      }
-    }
-  }
-
-  void GenArrayPackDecl_ObjectAPI(const StructDef &struct_def,
-                                  std::string *code_ptr,
-                                  std::vector<std::string> name_vec,
-                                  std::vector<int> array_length_vec) const {
+  void GenStructPackDecl_ObjectAPI(const StructDef &struct_def,
+                                   std::string *code_ptr,
+                                   const std::vector<FieldInfo> fields) const {
     auto &code = *code_ptr;
     for (auto it = struct_def.fields.vec.begin();
          it != struct_def.fields.vec.end(); ++it) {
       auto &field = **it;
       auto is_array = IsArray(field.value.type);
       const auto &field_type =
           is_array ? field.value.type.VectorType() : field.value.type;
-      if (!IsStruct(field_type)) {
-        auto tmp_name_vec = name_vec;
-        tmp_name_vec.push_back(field.name);
-        auto tmp_array_length_vec = array_length_vec;
-        if (is_array) {
-          tmp_array_length_vec.push_back(field_type.fixed_length);
+      auto tmp_fields = fields;
+      FieldInfo tmp_field = {
+        field.name,
+        field_type.fixed_length,
+      };
+      tmp_fields.push_back(tmp_field);
+      if (field_type.struct_def != nullptr) {
+        GenStructPackDecl_ObjectAPI(*field_type.struct_def, code_ptr,
+                                    tmp_fields);
+      } else {
+        auto contains_array = false;
+        for (size_t i = 0; i < tmp_fields.size(); ++i) {
```

Comment:
here you need to go thru this array just to see if theres any `>0` elements, then below you need to explictly skip all `<=0` elements in every loop. Sounds to me `<=0` elements shouldn't be in this array in the first place, or in a different array.

Reply:
Looking over it, I think that would complicate the code further, especially when also adding in the `push_back()`/`pop_back()` change.

Reply:
I think the code is already complicated for what it does, so I'd like to clean it up with less special cases and superfluous code.

Reply:
Ok, is it better now?